publication/models/account_analytic_invoice_line.py

```diff
@@ -28,3 +28,87 @@ def action_distribution_list(self):
         action['view_mode'] = 'form'
         action['target'] = 'current'
         return action
+
+    @api.model
+    def create(self, vals):
+        result = super(AccountAnalyticInvoiceLine, self).create(vals)
+        if result.publication:
+            self._update_publication_list()
+        return result
+
+    @api.multi
+    def write(self, vals):
+        p_product = self.env['product.product']
+        d_list = self.env['distribution.list']
+        product_id = vals.get('product_id')
+        # if the user set changed the product to a non publication product
+        # we need to find and delete the corresponding publication list
+        if product_id and not p_product.browse(
+                product_id).publication:
+            contract_partner_ids = self.mapped(
+                'analytic_account_id').mapped('partner_id').ids
+            line_partner_ids = self.mapped('partner_id').ids
+            product_ids = self.mapped('product_id').ids
+            d_list.search([
+                ('contract_partner_id', 'in', contract_partner_ids),
+                ('partner_id', 'in', line_partner_ids),
+                ('product_id', 'in', product_ids),
+            ]).unlink()
```

Comment:
> this will unlink any combination of contract_partner_id, partner_id, product_id in your set, not what you want for n > 1

There should not be more than one records that satisfy this query for the same combination. If that is the case then I think maybe this is wrong on a conceptual level and these three might very well be used as primary keys rather than hoping around.

> Further, you shouldn't unlink anything at all here, but remove the amount of copies now not contracted any more from the contract partner distribution list entry. If there are too little copies assigned there, it's an error

As I understand this request conflicts with https://github.com/NL66278/crm/pull/7#issuecomment-463566646

Reply:
1) consider partners A, B and products P1,  and P2. We have a contract line (A, A, P1), (A, A, P2), (B, B, P2). Now we change the product of (A, A, P1) and (B, B, P2) simultaneously. What happens?
2) it doesn't. This is about when *all* contract lines for a partner/product combination are deleted/ended, but nothing here checks for this, the code at hand only looks at the currently changed line(s)

Reply:
1)
> We have a contract line (A, A, P1), (A, A, P2), (B, B, P2)

By this notation I get you mean that we have three different contract lines. 
All the lines that are created on a contract will have the same partner_id. The same partner as the contract. The partner of a contract line _cannot_ be different from the partner on the contract.

Also, https://github.com/OCA/OCB/blob/10.0/odoo/fields.py#L2234 which means that the write will run once per record that was changed.

2) Still, throwing an error here will prohibit users from deleting contract lines when someone has changed the quantity on the distribution lists to something lower.

I will also need an answer to:
> There should not be more than one records that satisfy this query for the same combination. If that is the case then I think maybe this is wrong on a conceptual level and these three might very well be used as primary keys rather than hoping around.

Reply:
1) you misunderstand your own code: I'm talking about the `distribution.list` records you delete, you will also delete (A, A, P2) entries. And of course you won't have such a situation in the many2one on an analytic contract, but there might be a flat list of contract lines somewhere in the UI, or code that selects some contract lines based on whatever criteria that yield a situation as described. Your code must be resilient to all those possibilities. And what you link happens to be the case if the write comes from a one2many field as in the analytic contract form, and is by no means a guarantee for anything.
2) no it would not if you did the rest right, see the other comments and the specification
3) read the code of `publication` until you actually understand it which quite obviously is not the case currently, then you'll see what you claim here is wrong bordering nonsense.